### ISSUE TITLE
[Feature request]: Advanced pdf features

### ISSUE THREAD

**Issue by CodeDoctorDE** [LinwoodDev/Butterfly#327]
### Is your feature request related to a problem? Please describe

PDFs are a big topic for importing and exporting.
Featuring text support, ... would help people using the app in production

### Describe your feature request!

Add this

### Additional context

We need to create our own library for this.
https://github.com/DavBfr/dart_pdf does not support importing pdfs in the free tier.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cbk6028** [comment]
This is a good package. And I have some advices for loading pdf. There is a package  [pdfrx](https://github.com/espresso3389/pdfrx) backed by pdfium that support some advanced pdf features. Specially, so far Butterfly only imports pdf to images as canvas background. Butterfly's Renderer use canvas to draw backgroud which it is not better for rendering pdf clearly. Maybe put canvas on pdfrx viewer widget ？https://github.com/espresso3389/pdfrx?tab=readme-ov-file#additional-customizations. It provides Per-page Customization using Widget  and Per-page Customization using Canvas.

**Comment by CodeDoctorDE** [comment]
I'm usually against non-canvas renderers. My whole rendering system relies on the canvas and it allows saving performance. A widget would mean building an extra widget tree to display these.

My other idea was a "dynamic image" builder. It doesn't convert it on import, instead it saves the PDF and depending on the zoom level, it renders the image

**Comment by cbk6028** [comment]
I'm not sure meaning of "dynamic image" builder. Isn't this a pdf render?  I think pdfium directly render as an image depending on the zoom level. And I guess that you want to render pdf only by canvas. But comparing with pdfium, pdfium is a mature library. I think rendering by canvas may be a loss of performance.

**Comment by CodeDoctorDE** [comment]
Yeah, a real pdf renderer that renders an image and paints it on the canvas.
The image render part uses pdfium under the hood in the current version, so it doesn't make a difference . If we use widgets or canvas

**Comment by cbk6028** [comment]
Oh, I see.

**Comment by CodeDoctorDE** [comment]
I just looked at it and pdfrx seems like to have a canvas system inside it that I can use.
Putting it on my todo list

**Comment by longboat-dev** [comment]
Pdfrx use pdfium. You can create pdf by pdfium api directly. btw pdfrx may not bind all pdfium api with ffigen, you ffigen them.

**Comment by longboat-dev** [comment]
Cool, rust is good. But I think pdfium is a mature library which also supports wasm now.

**Comment by CodeDoctorDE** [comment]
Hmm, okay...
I'm not a big c++ pro but maybe I can do some things with ffigen and pdfium to get my own bindings to pdfium...

**Comment by longboat-dev** [comment]
https://github.com/ajrcarey/pdfium-render has rust bindings

**Comment by CodeDoctorDE** [comment]
This looks wonderful, I will try it

**Comment by CodeDoctorDE** [comment]
Added in 2.4.0-beta.3. I think "advanced" pdf features like highlighting or more will come very far in the future since the pdf library we use currently only allows this in a premium plan
